signature/src/lib.rs

```diff
@@ -1,17 +1,70 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
 //! Module for signature primitives.
 
-use crate::errors::PrimitivesError;
+#![cfg_attr(not(feature = "std"), no_std)]
+// Temporarily allow warning for nightly compilation with [`displaydoc`].
+#![allow(warnings)]
+#![deny(missing_docs)]
+#[cfg(test)]
+extern crate std;
+
+#[cfg(any(test, feature = "schnorr"))]
+#[macro_use]
+extern crate derivative;
+
+#[cfg(any(not(feature = "std"), target_has_atomic = "ptr"))]
+#[doc(hidden)]
+extern crate alloc;
+
 use ark_std::rand::{CryptoRng, RngCore};
 
+#[cfg(any(test, feature = "bls"))]
 pub mod bls_over_bls12381;
+#[cfg(any(test, feature = "bls"))]
 pub mod bls_over_bn254;
+pub mod constants;
+#[cfg(feature = "gadgets")]
+pub mod gadgets;
+#[cfg(any(test, feature = "schnorr"))]
 pub mod schnorr;
 
-pub use bls_over_bls12381::BLSSignatureScheme;
+use ark_std::{
+    format,
+    string::{String, ToString},
+};
+use blst::BLST_ERROR;
 use core::fmt::Debug;
-pub use schnorr::SchnorrSignatureScheme;
+use displaydoc::Display;
 use serde::{Deserialize, Serialize};
 use zeroize::Zeroize;
+
+/// Signature error type
+#[derive(Debug, Display, Eq, PartialEq)]
+pub enum SignatureError {
+    /// Bad parameter in function call, {0}
+    ParameterError(String),
+    /// Verification failed, {0}
+    VerificationError(String),
+}
+
```

Comment:
can we add a TODO to reconsider this choice? 

most of the time I want to use `VerificationError` to enforce proper handling of those err, but this enum helps escape this guarantee.